client/src/connection/rest/RestContentAdapter.ts

```diff
@@ -187,30 +193,96 @@ class RestContentAdapter implements ContentAdapter {
       return "";
     }
 
-    const filePathParts = [];
-    let currentContentItem: Pick<ContentItem, "parentFolderUri" | "name"> =
-      item;
-    if (!folderPathOnly) {
-      filePathParts.push(currentContentItem.name);
+    // Use different cache keys for folder paths vs full paths
+    const cacheKey = folderPathOnly ? `${item.id}_folder` : item.id;
+
+    if (this.pathCache.has(cacheKey)) {
+      return this.pathCache.get(cacheKey)!;
+    }
+
+    try {
+      const path = await this.calculatePathOfItem(item, folderPathOnly);
+      this.pathCache.set(cacheKey, path);
+      return path;
+    } catch {
+      return "";
     }
+  }
+
+  private async calculatePathOfItem(
+    item: ContentItem,
+    folderPathOnly?: boolean,
+  ): Promise<string> {
+    // If folderPathOnly=true, we want the path of the parent folder
+    if (folderPathOnly) {
+      if (!item.parentFolderUri) {
+        return "";
+      }
+
+      const cachedParentPath = this.pathCache.get(item.parentFolderUri);
+      if (cachedParentPath) {
+        return cachedParentPath;
+      }
+
+      try {
+        const { data: parentData } = await this.connection.get(
+          item.parentFolderUri,
+        );
+        const parentPath = await this.calculatePathOfItem(parentData, false); // Get parent's full path
+        this.pathCache.set(item.parentFolderUri, parentPath);
+        return parentPath;
+      } catch {
+        return "";
+      }
+    }
+    const filePathParts = [];
+    let currentContentItem: Pick<
+      ContentItem,
+      "parentFolderUri" | "name" | "id"
+    > = item;
```

Comment:
Could this just be...
```suggestion
    let currentContentItem: ContentItem = item;
```